File: study/ts/script/data/GemstoneData.ts
import { BaseData } from "./BaseData";
import { MessageIDConst } from "../const/MessageIDConst";
import { handler } from "../utils/handler";
import { G_NetworkManager, G_ConfigLoader } from "../init";
import { TypeConvertHelper } from "../utils/TypeConvertHelper";
import { ConfigNameConst } from "../const/ConfigNameConst";

export interface GemstoneData {
    getGemstones(): Object
    setGemstones(value: Object): void
    getLastGemstones(): Object
}
interface GemstoneUnitData {
    getId(): number
    setId(value: number): void
    getLastId(): number
    getNum(): number
    setNum(value: number): void
    getLastNum(): number
    getType(): number
    setType(value: number): void
    getLastType(): number
    getConfig(): Object
    setConfig(value: Object): void
    getLastConfig(): Object
}
let schema = {};
schema['id'] = [
    'number',
    0
];
schema['num'] = [
    'number',
    0
];
schema['type'] = [
    'number',
    0
];
schema['config'] = [
    'object',
    {}
];
class GemstoneUnitData extends BaseData {
    public static schema = schema;
    constructor() {
        super()
    }

    initData(data): void {
        this.setId(data.id);
        this.setNum(data.num);
        this.setType(TypeConvertHelper.TYPE_GEMSTONE);
        var info = G_ConfigLoader.getConfig(ConfigNameConst.GEMSTONE).get(data.id);
        this.setConfig(info);
    }

}
schema = {};
schema['gemstones'] = [
    'object',
    {}
];
export class GemstoneData extends BaseData {
    public static schema = schema;
    constructor() {
        super()
        this._gemstoneList = {};
        this._recvGetGemstone = G_NetworkManager.add(MessageIDConst.ID_S2C_GetGemstone, this._s2cGetGemstone.bind(this));
    }

    private _recvGetGemstone;
    private _gemstoneList;

    clear() {
        this._recvGetGemstone.remove();
        this._recvGetGemstone = null;
    }
    reset() {
        this._gemstoneList = {};
    }
    _setGemstoneData(data) {
        var unitData = new GemstoneUnitData();
        unitData.initData(data);
        this._gemstoneList['k_' + (data.id)] = unitData;
    }
    _s2cGetGemstone(id, message) {
        this._gemstoneList = {};
        var gemstones = message['gemstones'] || {};
        for (var i in gemstones) {
            var data = gemstones[i];
            this._setGemstoneData(data);
        }
    }
    updateData(datas) {
        if (datas == null || typeof (datas) != 'object') {
            return;
        }
        if (this._gemstoneList == null) {
            return;
        }
        for (var i = 0; i < datas.length; i++) {
            this._setGemstoneData(datas[i]);
        }
    }
    insertData(datas) {
        if (datas == null || typeof (datas) != 'object') {
            return;
        }
        if (this._gemstoneList == null) {
            return;
        }
        for (var i = 0; i < datas.length; i++) {
            this._setGemstoneData(datas[i]);
        }
    }
    deleteData(datas) {
        if (datas == null || typeof (datas) != 'object') {
            return;
        }
        if (this._gemstoneList == null) {
            return;
        }
        for (var i = 0; i < datas.length; i++) {
            var id = datas[i];
            this._gemstoneList['k_' + (id)] = null;
        }
    }
    getUnitDataWithId(id) {
        var data = this._gemstoneList['k_' + (id)];
        return data;
    }
    getGemstonesData(sortFuncType ?) {
        var sortFun;
        if (sortFuncType && sortFuncType == 1) {
            sortFun = function (a, b) {
                return b.getConfig().sorting - a.getConfig().sorting;
            };
        } else {
            sortFun = function (a, b) {
                return a.getConfig().sorting - b.getConfig().sorting;
            };
        }
        var result: any = [];
        for (var k in this._gemstoneList) {
            var unit = this._gemstoneList[k];
            unit && result.push(unit);
        }
        result.sort(sortFun);
        return result;
    }

}

File: study/lua/src/app/data/GemstoneData.lua
-- Author: liangxu
-- Date:2017-10-18 16:48:19
-- Describle：宝石数据

local BaseData = require("app.data.BaseData")
local GemstoneData = class("GemstoneData", BaseData)

--宝石单元数据
local GemstoneUnitData = class("GemstoneUnitData", BaseData)
local schema = {}
schema["id"] = {"number", 0}
schema["num"] = {"number", 0}
schema["type"] = {"number", 0}
schema["config"] = {"table", {}}
GemstoneUnitData.schema = schema

function GemstoneUnitData:ctor(properties)
	GemstoneUnitData.super.ctor(self, properties)
end

function GemstoneUnitData:initData(data)
	self:setId(data.id)
	self:setNum(data.num)
	local TypeConvertHelper = require("app.utils.TypeConvertHelper")
	self:setType(TypeConvertHelper.TYPE_GEMSTONE)

	local info = require("app.config.gemstone").get(data.id)
    assert(info,"gemstone config can't find id = "..tostring(data.id))
	self:setConfig(info)
end


--宝物数据
local GemstoneData = class("GemstoneData", BaseData)
local schema = {}
schema["gemstones"] = {"table", {}}
GemstoneData.schema = schema

function GemstoneData:ctor(properties)
	GemstoneData.super.ctor(self, properties)
	self._gemstoneList = {}
	self._recvGetGemstone = G_NetworkManager:add(MessageIDConst.ID_S2C_GetGemstone, handler(self, self._s2cGetGemstone))
end

function GemstoneData:clear()
	self._recvGetGemstone:remove()
	self._recvGetGemstone = nil
end

function GemstoneData:reset()
	self._gemstoneList = {}
end

function GemstoneData:_setGemstoneData(data)
    local unitData = GemstoneUnitData.new()
    unitData:initData(data)
    self._gemstoneList["k_"..tostring(data.id)] = unitData
end

--
function GemstoneData:_s2cGetGemstone(id, message)
	self._gemstoneList = {}
	local gemstones = rawget(message, "gemstones") or {}
	for i, data in ipairs(gemstones) do
		self:_setGemstoneData(data)
	end
end

function GemstoneData:updateData(datas)
	if datas == nil or type(datas) ~= "table" then
        return
    end
    if self._gemstoneList == nil then
        return
    end
	for i = 1, #datas do
		self:_setGemstoneData(datas[i])
	end
end

function GemstoneData:insertData(datas)
	if datas == nil or type(datas) ~= "table" then
        return
    end
    if self._gemstoneList == nil then
        return
    end

	for i = 1, #datas do
		self:_setGemstoneData(datas[i])
	end
end

function GemstoneData:deleteData(datas)
	if datas == nil or type(datas) ~= "table" then
        return
    end
    if self._gemstoneList == nil then
        return
    end

	for i = 1, #datas do
		local id = datas[i]
		self._gemstoneList["k_"..tostring(id)] = nil
	end
end

--
function GemstoneData:getUnitDataWithId(id)
	local data = self._gemstoneList["k_"..tostring(id)]
	return data
end

--sortFuncType 排序规则  1 按品质低-->高 nil 按品质高-->低
function GemstoneData:getGemstonesData(sortFuncType)
	local sortFun
	if sortFuncType and sortFuncType == 1 then
		sortFun = function(a, b)
			return a:getConfig().sorting > b:getConfig().sorting
		end
	else
		sortFun = function(a, b)
			return a:getConfig().sorting < b:getConfig().sorting
		end
	end

	local result = {}
	for k, unit in pairs(self._gemstoneList) do
		table.insert(result, unit)
	end
	table.sort(result, sortFun)

	return result
end

return GemstoneData
